common/src/memory_dump.rs

```diff
@@ -0,0 +1,98 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use gear_core::memory::{GearPage, PageBuf, PageU32Size};
+use path_clean::PathClean;
+use serde::{Deserialize, Serialize};
+use sp_core::{Decode, Encode};
+use std::{env, fs, path::Path};
+
+#[derive(Serialize, Deserialize)]
+pub struct MemoryPageDump {
+    page: u32,
+    data: Option<String>,
+}
+
+impl MemoryPageDump {
+    pub fn new(page_number: GearPage, page_data: PageBuf) -> MemoryPageDump {
+        let mut data_vec = vec![];
+        page_data.encode_to(&mut data_vec);
+        let data = if data_vec.iter().all(|&byte| byte == 0) {
+            None
+        } else {
+            Some(hex::encode(data_vec))
+        };
+        MemoryPageDump {
+            page: page_number.raw(),
+            data,
+        }
+    }
+
+    pub fn into_gear_page(self) -> (GearPage, PageBuf) {
+        let page_buf = if let Some(page_hex) = self.data {
+            let data = hex::decode(page_hex).expect("Unexpected memory page data encoding");
+            PageBuf::decode(&mut &*data).expect("Invalid PageBuf data found")
+        } else {
+            PageBuf::new_zeroed()
+        };
+        (
+            GearPage::new(self.page)
+                .unwrap_or_else(|_| panic!("Couldn't decode GearPage from u32: {}", self.page)),
+            page_buf,
+        )
+    }
+}
+
+#[derive(Serialize, Deserialize, Default)]
+pub struct ProgramMemoryDump {
+    pub balance: u128,
+    pub reserved_balance: u128,
+    pub pages: Vec<MemoryPageDump>,
+}
+
+impl ProgramMemoryDump {
+    pub fn save_to_file<P: AsRef<Path>>(&self, path: P) {
```

Comment:
`impl` is more readable for single generic.
```suggestion
    pub fn save_to_file(&self, path: impl AsRef<Path>) {
```